### ISSUE TITLE
[FEATURE] NFS support (RWX) GA

### ISSUE THREAD

**Issue by yasker** [longhorn/longhorn#2293]
We want to improve our NFS solution to be highly available, so one node down (especially the node running the NFS server) will result in minimal downtime of the NFS server and near-zero interruption for the NFS client.

One note: since the NFS server uses `sync` mode by default, all the data are written by the client will be written on the underlying block device before the request was acked.

Ref: https://www.avidandrew.com/understanding-nfs-caching.html (Server-side caching)

### Tasks: 
- [x] https://github.com/longhorn/longhorn/pull/4299
- [x] https://github.com/longhorn/longhorn/issues/4351
- [x] dedicated recovery backend: [longhorn-manager side](https://github.com/longhorn/longhorn-manager/pull/1464)
- [x] dedicated recovery backend: [nfs-ganesha side](https://github.com/longhorn/nfs-ganesha/pull/2)
- [x] nfs server grace priod, lease time and hostname issues (https://github.com/longhorn/longhorn-share-manager/pull/36)
- [x] longhorn-manager client mount options (https://github.com/longhorn/longhorn-manager/pull/1464)

**Comment by derekbit** [comment]
**Investigate two scenarios**

**1. node, which has the volume and the share-manager pod, is not crashed or shut down, but the share-manager pod is stopped unexpectedly**
    The  share-manager pod  is restart on the same node, and then everything works well again.

**2. node, which has the volume and the share-manager pod, is crashed or suddenly shut down**
    The volume and share-manager are re-created on another node.
    The NFS4 client try to resend SEQUENCE to the new server but encounters timeout error for 5-6 mins. [client-rpc.log](https://github.com/longhorn/longhorn/files/8887142/client-rpc.log)
    The connect is stuck somewhere, and the SEQUENCE is not received at the server side. 
    After 5-6 mins, the old connection is cleaned up successfully, and the NFS4 connection becomes normal.

**Comment by derekbit** [comment]
**OS: ubuntu 22.04 (kernel: 5.15.0)**

The incomplete IO operation is retried in the second case in the above comment after the volume and share-manager pod are re-created on another node. The NFS-HA is achieved.

The iptable rules contain more sophisticated traffic rules between the app pod share-manager pod in this environment. These rules are not seen in the above one. We need to investigate iptables more thoroughly.

**Comment by voarsh2** [comment]
Could making the share-manager run thru a headless service or ClusterIP, with replica's? K8 design is to have more than one instance running and share-manager is not using replica's.....

**Comment by derekbit** [comment]
The `Lease_Lifetime` of the NFS server is 60 seconds. The value needs to be decreased but also investigate if there is any  downside.

https://www.suse.com/support/kb/doc/?id=000019374
https://github.com/longhorn/longhorn-share-manager/blob/5b57b596e192d758f3049142a7ced47c0b0dae03/pkg/server/nfs/nfs_server.go#L37

**Comment by derekbit** [comment]
Assume the failover time should be smaller than 30 seconds.
There are several factors we need to consider in order to achieve the NFS-HA.

- Default kubernetes `node-monitor-grace-period` is 40 seconds. It means the node status is set to`down`/`not ready` when the node is down for 40 seconds. Thus, the value should be set to be smaller than 30 seconds, e.g 12 seconds.
    - The downside is that a temporary failure of network for a small period would result in pod scheduling.
 
- Need a `share-manager` controller to deploy the share-manager image on each Longhorn node. Then, the share-manager image can be created on another node quickly if the original node is down.

- The `Lease_Lifetime` of the NFS server is 60 seconds. The value also needs to be smaller.

---

I set
- `node-monitor-grace-period` to 12 seconds
-  pre-pulled the share-manager image on each node
-  set the Lease_Lifetime to 5 seconds. 

Without leveraging the live migration concept (one active engine/one passive engine), if the node, where the share-manager pod, is on is down, then the rwx volume is detached and attached to another node and the share-manager pod is re-created to another node. The average failover period is 23 seconds.

@joshimoo @shuo-wu @innobead 
Do you think if this available in our first implementation for NFS-HA?

**Comment by innobead** [comment]
> * Default kubernetes `node-monitor-grace-period` is 40 seconds. It means the node status is set to`down`/`not ready` when the node is down for 40 seconds. Thus, the value should be set to be smaller than 30 seconds, e.g 12 seconds.
>   
>   * The downside is that a temporary failure of network for a small period would result in pod scheduling.

I don't think we should have a best practice for this setting, but rather describe the HA will respect or be impacted by this period setting. 

> * Need a `share-manager` controller to deploy the share-manager image on each Longhorn node. Then, the share-manager image can be created on another node quickly if the original node is down.

Sounds good to me and it seems general to be able to reuse for other on-demand pods like backing image. 

> * The `Lease_Lifetime` of the NFS server is 60 seconds. The value also needs to be smaller.

Do we have any reference value from other HA examples? 5 seconds seems very aggressive, any possible downside?

> I set
> 
> * `node-monitor-grace-period` to 12 seconds
> * pre-pulled the share-manager image on each node
> * set the Lease_Lifetime to 5 seconds.


> Without leveraging the live migration concept (one active engine/one passive engine), if the node, where the share-manager pod, is on is down, then the rwx volume is detached and attached to another node and the share-manager pod is re-created to another node. The average failover period is 23 seconds.
> 
> Although leverage the live migration, there are passive engine/replicas, the failover time is quite similar.

Question: when using the live migration, what extra cost we will have? extra engine process only or extra replicas as well?

Also, if `node-monitor-grace-period` cant be optimized, will adopting the live migration help?

**Comment by derekbit** [comment]
> > * Default kubernetes `node-monitor-grace-period` is 40 seconds. It means the node status is set to`down`/`not ready` when the node is down for 40 seconds. Thus, the value should be set to be smaller than 30 seconds, e.g 12 seconds.
> >   
> >   * The downside is that a temporary failure of network for a small period would result in pod scheduling.
> 
> I don't think we should have a best practice for this setting, but rather describe the HA will respect this period setting or be impacted by it.

OK. We can address the setting in the document.

> 
> > * Need a `share-manager` controller to deploy the share-manager image on each Longhorn node. Then, the share-manager image can be created on another node quickly if the original node is down.
> 
> Sounds good to me and it seems general to be able to reuse for other on-demand pods like backing image. (can create another task for that if needed)
> 
> > * The `Lease_Lifetime` of the NFS server is 60 seconds. The value also needs to be smaller.
> 
> Do we have any reference value from other HA examples? 5 seconds seems very aggressive, any possible downside?

A low `Lease_Lifetime` value might result in loss of leases and locks when the network condition is not stable. I think this means the mounted filesystem becomes unstable under a bad networking condition.
.
SUSE doc recommends the lease value should be larger than 30 seconds (default: 60 seconds) for safety.
https://www.suse.com/support/kb/doc/?id=000019414

When the new nfs server starts, the nfs client will stall for a while (roughly equal to `Lease_Lifetime`).

> 
> > I set
> > 
> > * `node-monitor-grace-period` to 12 seconds
> > * pre-pulled the share-manager image on each node
> > * set the Lease_Lifetime to 5 seconds.
> 
> > Without leveraging the live migration concept (one active engine/one passive engine), if the node, where the share-manager pod, is on is down, then the rwx volume is detached and attached to another node and the share-manager pod is re-created to another node. The average failover period is 23 seconds.
> > Although leverage the live migration, there are passive engine/replicas, the failover time is quite similar.
> 
> Question: when using the live migration, what extra cost we will have? extra engine process only or extra replicas as well?

If we leverage the live migration concept, we need an extra engine and extra replicas. In addition, we need to handle many engine and replica operations, e.g. deletion, replenish, rebuild and etc. These operations are not supported in original live migration logic.

> 
> Also, if `node-monitor-grace-period` cant be optimized, will adopting the live migration help?

I think no, because the controller needs to rely on the node status. `node-monitor-grace-period` introduces 40 seconds delay by default, and NFS default lease lifetime introduces 60 seconds. In addition, the share-manager pod nfs server start need extra time consumption. The total failover time is already larger than 60 seconds.

**Comment by joshimoo** [comment]
@derekbit I agree that it would be simpler to start without active/passive (i.e. no second engine) and just rely on the share-manager pod being started on a different node.

Grace Period should be at least 2x Lease_Lifetime so that there is a chance to reclaim the locks.
Since currently we don't store the metadata persistently we cannot do lock reclaim, therefore after the NFS server starts the clients will wait till Grace Period, if we supported the metadata volume. Lock reclaim would work and once all locks are reclaimed, the grace period can be ended early.
https://github.com/longhorn/longhorn-share-manager/blob/5b57b596e192d758f3049142a7ced47c0b0dae03/pkg/server/nfs/nfs_server.go#L37-L39

**Comment by innobead** [comment]
> @derekbit I agree that it would be simpler to start without active/passive (i.e. no second engine) and just rely on the share-manager pod being started on a different node.
> 
> Grace Period should be at least 2x Lease_Lifetime so that there is a chance to reclaim the locks. Since currently we don't store the metadata persistently we cannot do lock reclaim, therefore after the NFS server starts the clients will wait till Grace Period, if we supported the metadata volume. Lock reclaim would work and once all locks are reclaimed, the grace period can be ended early. [longhorn/longhorn-share-manager@`5b57b59`/pkg/server/nfs/nfs_server.go#L37-L39](https://github.com/longhorn/longhorn-share-manager/blob/5b57b596e192d758f3049142a7ced47c0b0dae03/pkg/server/nfs/nfs_server.go#L37-L39)

It's interesting. @joshimoo could you elaborate more about `the metadata volume`, is this what we can control or actually at ganesha part?

**Comment by derekbit** [comment]
A NFS recovery backend records the client id when a client connects the NFS server, and it will be utilized for the recovery process after the server is crashed and comes back.

Client ID is recorded in a NFS recovery backend when a client connects to the NFS server, and it will be used to recovery the connection after the server crashes and then comes back.

Longhorn share-manager's NFS server uses the `fs_ng` as the recovery backend. That means all Longhorn nodes must be able to access the same recovery backend (filesystem). Otherwise, the recovery might fail.

In order to continue using `fs_ng`, I've investigated several approaches

- ConfigMap: The ConfigMaps are always mounted read-only volumes, so it doesn't fit our needs. 
- NFS export folder from an external machine: The method will make the RWX volume feature depends on external components.
- NFS export folder from the api-server: If the apiserver is high available, we need to make the NFS export folder high available.

`fs_ng` looks not feasible in Longhorn system. A possible approach is to implement a customized backend such as`etcd_kv` recovery backend for our application.

For example, there is a `rados_kv` backend in nfs_ganesha.
Ref. https://github.com/nfs-ganesha/nfs-ganesha/blob/V3.4/src/SAL/recovery/recovery_rados_kv.c

@joshimoo @innobead WDYT?

**Comment by joshimoo** [comment]
@derekbit correct implementing our own recovery backend either based on etcd or via an api on longhorn/api server extension, seems like a good idea once we have all the other nfs moving parts working.

You can continue with the CIFS mount for the POC test, to ensure that we addressed all the other issues.

Also based on our slack communication, if you set the UTS namespace (i.e. hostname) before starting the longhorn-share-manager application / ganesha, that might help with the file handle issues, due to different uts scope when starting on a different node.

MORE reference material here:
- https://github.com/ClusterLabs/resource-agents/issues/1644
- https://access.redhat.com/solutions/1179643

**Comment by innobead** [comment]
@derekbit for the recover backend, how will it be integrated into ganesah used by our share manager? Is it just an external binary able to hook into the server at runtime or does it need to be statically compiled into the ganesha executable at compile time?

Yeah, agreed with that to use in-cluster resource like etcd_kv you named, but it should be k8s_api backend instead. I don't think we want any etcd operations directly but rather use the K8s resource instead.

**Comment by derekbit** [comment]
@innobead It needs to be compiled into the nfs-ganesha executable at compile time.
The nfs-server is running in the share-mamager pod, so I'm still thinking how to access the in-cluster resource. Any suggestion?

**Comment by derekbit** [comment]
- The hostname of the share-manager pod is already the pod name and won't be changed after failovering to another node.
- CIFS does not support symbolic link operation, so I replaced it with an external folder as a recovery backend.

Using the test method in https://github.com/ClusterLabs/resource-agents/issues/1644, the io error happened after failover caused by `NFS: __nfs4_reclaim_open_state: Lock reclaim failed!`.

But the `_nfs4_proc_exchange_id: server_scope mismatch detected` is not seen in the client side's dmesg log as described in https://github.com/ClusterLabs/resource-agents/issues/1644. It means current failure's root cause is not a server_scope issue.

**Comment by derekbit** [comment]
Update the work on https://github.com/longhorn/longhorn/issues/2293#issuecomment-1193494986:

`fs_ng` recovery backend stores the client information as the format  `::ffff:${client IPv4}-(31:Linux NFSv4.1 ${node name})`.

My environemt is k3s with flannel CNI.
 
- When the client and nfs server are on different two nodes, the `${client IPv4}` is the IP address of `flannel.1` interface of the client's host.

- When the client and nfs server are on the same node, the `${client IPv4}` is the IP address of `cni0` interface of the client's host.

Thus, if the client is on `node 1` and server is on `node 2`
- The failover succeeds after the server failovers to `node 3`.
- The failover fails after the server failovers to `node 1`. This is because the server uses a different client IP, `cni0`'s IP, to search the record in the recovery backend.

To address this issue, we can hack the nfs-ganesha and don't need to remember the client IP in the recovery backend such as https://github.com/derekbit/nfs-ganesha/commit/47c25fb6afb7f31c99ac50eb7835817bf4c746d0. This idea works and the failover works under the test https://github.com/ClusterLabs/resource-agents/issues/1644#issue-879309588.

cc @joshimoo

**Comment by joshimoo** [comment]
@derekbit great job on the investigation and testing :)
I wonder though if the connection was via `cni0` before the failure, what happens to the connection after the failover?
Since it would need to reroute via `flannel.1` does the nfs client by default have the appropriate retry logic for this?

**Comment by joshimoo** [comment]
@derekbit while working on this, can you also update the underlying ganesha version :)
It's a good way to get comfortable with it, there is a custom patch applied to it.
https://github.com/longhorn/nfs-ganesha/commit/1c27c2b8667b849c0b1b1d8fb718a736ddec393f

The patch changes just a little bit with the newer versions, let me know if you have issues with that.
Afterwards need to create a new release in https://github.com/longhorn/nfs-ganesha then update the reference in the build script here: https://github.com/longhorn/longhorn-share-manager/commit/9e021acdf39ec41e60b99564db1c504f8eff4683

**Comment by derekbit** [comment]
> longhorn/longhorn-share-manager@9e021ac

No problem. My nfs-ganesha is based on https://github.com/longhorn/nfs-ganesha/commit/1c27c2b8667b849c0b1b1d8fb718a736ddec393f.

**Comment by derekbit** [comment]
> @derekbit great job on the investigation and testing :)
> 
> I wonder though if the connection was via `cni0` before the failure, what happens to the connection after the failover? Since it would need to reroute via `flannel.1` does the nfs client by default have the appropriate retry logic for this?
> 
> We should probably also consider a different approach to uniquely identifying the clients, unless CID is already sufficient, don't know what that looks like of the top of my head :)

@joshimoo 

IIUC, if the client and the server on the same node, the reroute via `flannel.1` is not required, so the server will record the client using the `cni0` IP. The route path is `client->cni0->server`.

However, if they are on different nodes, the reroute is inevitable, and the `flannedl.1` is recorded. The route path becomes `client->cni0->flannel.1->server`.

If the server goes to the client's node after failover, the client IP becomes `cni0` IP. I didn't see any retry using the `flannel.1` interface from log.


> We should probably also consider a different approach to uniquely identifying the clients, unless CID is already sufficient, don't know what that looks like of the top of my head :)

Agree. But I cannot find a good identification for now. I'm going to check how rados backend and other backends handle this part first.

**Comment by derekbit** [comment]
Here is the PoC setup and result

---
**Setup**

3 worker nodes for the Longhorn cluster

- Attach 1 RWO volume to node 1
- Attach 2 RWO volumes to node 2
- Attach 2 RWO volumes to node 3

Then, create 1 RWX volume. Then, Run an app pod with the RWX volume on each worker node.
Execute the command in each app pod
```
( exec 7<>/data/testfile-${i}; flock -x 7; while date | dd conv=fsync >&7 ; do sleep 1; done )
```
where #{i} is the node number.

nfs server's lease time is 60 seconds and the grace period is 90 seconds.

---
**Result**

share-manager pod (nfs-server) is at node-1 in the beginning. Then, it failovers to node-3 after I shut the node-1 down.
The clients on node-2 and node-3's app pods can reclaim the locks in 10 seconds and exit the grace period early. 

Although the client on node-1's app pod died and one of the locks can not be reclaimrf, the lock reclaim can still be finished early.

cc @joshimoo

**Comment by derekbit** [comment]
Proceeding with the PoC tests

1. Turn into the deployment into daemonset in https://github.com/longhorn/longhorn/blob/master/examples/rwx/rwx-nginx-deployment.yaml and disable `Automatically Delete Workload Pod when The Volume Is Detached Unexpectedly`. Then, deploy the daemonset with a RWX volume.

    After shutting down the nfs server's node, the clients didn't run into the stale handle errors after the failover. The locks were reclaimed early as well.


2. Byte-range file locking

    I also wrote a file range locking test ([range_locking.txt](https://github.com/longhorn/longhorn/files/9208112/range_locking.txt)).
    The client in each app pod locked the same file's differenet range.
    Although one of clients is down and the its lock cannot be reclaimed, the lock reclaim process can be finished early and the clients continue the tasks after the server's failover without IO or stale handle errors.

cc @joshimoo

**Comment by innobead** [comment]
See if it's possible to add e2e test cases in the current daily regression pipeline or need a new pipeline for this. @derekbit WDYT?

cc @longhorn/qa

**Comment by derekbit** [comment]
Yes, can add e2e test cases.

**Comment by chriscchien** [comment]
**Verified in longhorn-manager(`259cf1`) with test [steps](https://github.com/longhorn/longhorn/pull/4299/files?short_path=ede263f#diff-ede263f3ab5063ec016e12aa93c463c9dd10df20965377d5f64859ff1c1a3085)
Result Pass**

1. Create 1 RWX volume and then run an app pod with the RWX volume on each worker node. Execute the command `exec 7<>/data/testfile-${i}; flock -x 7; while date | dd conv=fsync >&7 ; do sleep 1; done` (${i} is the node number) in each app pod. Then turn off the node where share-manager is running trigger failover

   - After share-manager failover to another node ready , the I/O in exist RWX pod can continue working properly without error
   - And data integrity checked OK by command `md5sum` with an exist file in mount point

2. Turn the [deployment](https://github.com/longhorn/longhorn/blob/master/examples/rwx/rwx-nginx-deployment.yaml) into a daemonset and disable `Automatically Delete Workload Pod when The Volume Is Detached Unexpectedly`. Then, deploy the daemonset with a RWX volume. Then turn off the node where share-manager is running trigger failover

   - After share-manager failover to another node ready , the I/O in exist RWX pod can continue working properly without error
   - And data integrity checked OK by command `md5sum` with an exist file in mount point

3. Byte-range file locking by run each client [range_locking.c](https://github.com/longhorn/longhorn/files/9208112/range_locking.txt)(use gcc to compile) in each app pod locks a different range of the same file. Afterwards, it writes data repeatedly into the file. Then turn off the node where share-manager is running trigger failover
  
   - The clients continue the tasks after the server's failover without IO or stale handle errors
   - And data integrity checked OK by command `md5sum` with an exist file in mount point

Currently I can not find a proper way to test `Lock reclaim process can be finished earlier than the 90-seconds grace period.` 
I will test lock reclaim once have a reproducible steps. thank you

cc @derekbit

**Comment by innobead** [comment]
Need to update the doc, https://longhorn.io/docs/1.4.0/advanced-resources/rwx-workloads/

**Comment by chriscchien** [comment]
Close this ticket as test stpes [verified](https://github.com/longhorn/longhorn/issues/2293#issuecomment-1301743625) and document merged
